### PR TITLE
Add Ben-Or Byzantine agreement algorithm

### PR DESCRIPTION
**Pull request by tmarzec**

### CODE REVIEW

**Review by krzysztof-turowski — CHANGES_REQUESTED** [review]
First round of suggestions - please synchronize PR with recent changes and consider how to align runBenOr with the code of runHumblet to make them as similar as possible.

I would think of it independently and we'll reach some agreement in the next round, I suppose.

File: Makefile [hunk 1]
```diff
@@ -26,6 +26,9 @@ sync_mst_example:
 sync_mis_example:
 	go run example/sync_mis_luby.go 20 0.25
 
+async_ben_or_example:
+	go run example/async_ben_or.go 6 1
```
[the remark this answers is not in the document]
Comment (tmarzec) [reply Makefile]: Done, https://github.com/tmarzec/distributed-framework/blob/91cac47f258419694da1fd8e4a564771a5eac642/Makefile#L29-L30

File: consensus/ben_or/ben-or.go [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+import (
+	"encoding/json"
+	"log"
+
+	"github.com/krzysztof-turowski/distributed-framework/lib"
+)
+type messageTypeBenOr int
+
+const (
+	msgDecided		= 1
```
[the remark this answers is not in the document]
Comment (tmarzec) [reply consensus/ben_or/ben-or.go]: Done.

File: consensus/ben_or/ben-or.go [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+	// nontrivial
+	for _, v := range vertices {
+		if starting := (*v).node.GetState()[0]; starting == firstBit {
+			log.Println("OK, decided value:", firstBit)
+			return
+		}
+	}
+
+	panic("The decided value differs from every starting value")
+}
```
[the remark this answers is not in the document]
Comment (tmarzec) [reply consensus/ben_or/ben-or.go]: Added.

File: lib/lib_nodes.go [hunk 4]
```diff
@@ -35,3 +36,11 @@ func getSynchronousChannels(n int) []chan []byte {
...
 	return channels
 }
+
+func getAsynchronousChannels(n int) []chan []byte {
+	channels := make([]chan []byte, n)
+	for i := range channels {
+		channels[i] = make(chan []byte)
+	}
+	return channels
+}
```
[the remark this answers is not in the document]
Comment (tmarzec) [reply lib/lib_nodes.go]: Added.

File: consensus/ben_or/ben-or.go [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+
+	// unanimous decision
+	for _, v := range vertices {
+		if decided := (*v).node.GetState()[1]; decided != firstBit {
+			panic("Decided values differ")
+		}
+	}
+	// nontrivial
+	for _, v := range vertices {
+		if starting := (*v).node.GetState()[0]; starting == firstBit {
```
[the remark this answers is not in the document]
Comment (tmarzec) [reply consensus/ben_or/ben-or.go]: It does, I've changed it now.

File: consensus/ben_or/ben-or.go [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+	node.FinishProcessing(true)
+}
+
+// validation
+func checkBenOr(vertices []*vertexBenOr) {
+	if len(vertices) == 0 {
+		log.Println("OK, length 0")
+		return
+	}
+	firstBit := (*vertices[0]).node.GetState()[1]
```
[the remark this answers is not in the document]
Comment (tmarzec) [reply consensus/ben_or/ben-or.go]: Done.

File: consensus/ben_or/ben-or.go [hunk 7]
```diff
@@ -0,0 +1,304 @@
+package ben_or
+
+import (
+	"encoding/json"
+	"log"
+
+	"github.com/krzysztof-turowski/distributed-framework/lib"
+)
+type messageTypeBenOr int
```
[the remark this answers is not in the document]
Comment (tmarzec) [reply consensus/ben_or/ben-or.go]: Done, I've changed it to: 
```golang
type messageTypeBenOr int


const (
	msgTypeOne messageTypeBenOr = iota
	msgTypeTwo
)
```

File: lib/lib_graph_weighted.go [hunk 8]
```diff
@@ -107,4 +107,4 @@ func BuildRandomConnectedWeightedGraph(n, m, maxWeight int, generator Generator)
 func BuildSynchronizedRandomConnectedWeightedGraph(n, m, maxWeight int, generator Generator) ([]WeightedGraphNode, Synchronizer) {
 	vertices, runner := BuildRandomConnectedWeightedGraph(n, m, maxWeight, generator)
 	return vertices, asSynchronizer(runner)
-}
+}
```
Comment (krzysztof-turowski) [inline lib/lib_graph_weighted.go:110]: Please add a blank line at the end

**Review by krzysztof-turowski — APPROVED** [review]
Thank you very much, aaside from the nice PR, the whole discussion was very fruitful and it clarified a lot for future development. Please run "go fmt" and add missing blank line (see comment) and you're good to go!

**Pull request merged**